src/Shared/Buffers.MemoryPool/PinnedBlockMemoryPool.cs

```diff
@@ -82,12 +134,94 @@ internal void Return(MemoryPoolBlock block)
             block.IsLeased = false;
 #endif
 
+        Interlocked.Increment(ref _returnCount);
+
         if (!_isDisposed)
         {
+            _metrics?.UpdateCurrentMemory(block.Memory.Length);
+            Interlocked.Add(ref _currentMemory, block.Memory.Length);
+
             _blocks.Enqueue(block);
         }
     }
 
+    public bool TryScheduleEviction(DateTimeOffset now)
+    {
+        if (now >= _nextEviction)
+        {
+            _nextEviction = now.Add(DefaultEvictionDelay);
+            ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: false);
+            return true;
+        }
+
+        return false;
+    }
+
+    void IThreadPoolWorkItem.Execute()
+    {
+        try
+        {
+            PerformEviction();
+        }
+        catch (Exception ex)
+        {
+            // Log the exception, but don't let it crash the thread pool
+            _logger?.LogError(ex, "Error while performing eviction in PinnedBlockMemoryPool.");
+        }
+    }
+
+    /// <summary>
+    /// Examines the current usage and activity of the memory pool and evicts a calculated number of unused memory blocks.
+    /// The eviction policy is adaptive: if the pool is underutilized or idle, more blocks are evicted;
+    /// if activity is high, fewer or no blocks are evicted.
+    /// Evicted blocks are removed from the pool and their memory is unrooted for garbage collection.
+    /// </summary>
+    internal void PerformEviction()
+    {
+        var currentCount = (uint)_blocks.Count;
+        var burstAmount = 0u;
+
+        var rentCount = _rentCount;
+        var returnCount = _returnCount;
+        _rentCount = 0;
+        _returnCount = 0;
+
+        // If any activity
+        if (rentCount + returnCount > 0)
+        {
+            // Trending less traffic
+            if (returnCount > rentCount)
+            {
+                // Remove the lower of 1% of the current blocks and 20% of the difference between rented and returned
+                burstAmount = Math.Min(currentCount / 100, (returnCount - rentCount) / 5);
+            }
+            // Traffic staying the same, try removing some blocks since we probably have excess
+            else if (returnCount == rentCount)
+            {
+                // Remove 1% of the current blocks (or at least 1)
+                burstAmount = Math.Max(1, currentCount / 100);
+            }
+            // else trending more traffic so we don't want to evict anything
+        }
+        // If no activity
+        else
+        {
+            // Remove 5% of the current blocks (or at least 10)
+            burstAmount = Math.Max(10, currentCount / 20);
+        }
+
+        // Remove from queue and let GC clean the memory up
+        while (burstAmount > 0 && _blocks.TryDequeue(out var block))
```

Comment:
I wonder if it'd be interesting to emit some metric if this `_blocks.TryDequeue` operation fails. It seems like this would be a good indicator that our heuristics aren't working as well as we'd like given how conservative we try to make the burst amount.

Reply:
Leave a comment on https://github.com/dotnet/aspnetcore/issues/61594? That could happen any time we are performing eviction and new traffic starts coming in, or if the pool was low and the heuristics didn't limit themselves to the size of the pool.